drake/solvers/Optimization.h

```diff
@@ -0,0 +1,323 @@
+    DecisionVariableView(const DecisionVariable& var, size_t start, size_t n) : var(var), start_index(start), length(n) {
+      assert(start+n<var.value.rows());
+    }
+
+    size_t index() const { return var.start_index + start_index; }
+    Eigen::VectorBlock<const Eigen::VectorXd,-1> value() const { return var.value.segment(start_index,length); }
+    size_t size() const { return length; }
+
+    const DecisionVariableView operator()(size_t i) const { assert(i<=length); return DecisionVariableView(var,start_index+i,1);}
+    const DecisionVariableView row(size_t i) const { assert(i<=length); return DecisionVariableView(var,start_index+i,1); }
+    const DecisionVariableView head(size_t n) const { assert(n<=length); return DecisionVariableView(var,start_index,n); }
+    const DecisionVariableView tail(size_t n) const { assert(n<=length); return DecisionVariableView(var,start_index+length-n,n); }
+    const DecisionVariableView segment(size_t start, size_t n) const { assert(start+n<=length); return DecisionVariableView(var,start_index+start,n); }
+
+    friend class OptimizationProblem;
+
+  private:
+    const DecisionVariable& var;
+    size_t start_index, length;
+  };
+
+  typedef std::list<DecisionVariableView> VariableList;
+
+  // Quick notes about this design:
+  // need to have lists of constraints (if not functions), so I want a (non-templated) base class.
+  // would be nice to have templated types for specialized functions.  Not strictly required.
+  // approach: let functions be templated, and use lambda closures/std::function to wrap them into the constraint classes.
+
+  /**
+   * @defgroup function_concept Function Concept
+   * @ingroup concepts
+   * @{
+   * @brief Describes a mapping from one Vector to another Vector y = f(x)
+   *
+   * @nbsp
+   *
+   * | Every model of this concept must implement |  |
+   * ---------------------|------------------------------------------------------------|
+   * | X::InputVector     | type for the input to the system, which models the Vector<ScalarType> concept |
+   * | X::OutputVector    | type for the output from the system, which models the Vector<ScalarType> concept |
+   * | template <ScalarType> OutputVector<ScalarType> operator()(const InputVector<ScalarType>& x) | the actual function |
+   * | InputOutputRelation getProperties() | |
+   * | size_t getNumInputs() | only required if the input vector is dynamically-sized |
+   * | size_t getNumOutputs() | only required if the output vector is dynamically-sized |
+   *
+   * (always try to label your methods with const if possible)
+   *
+   * @}
+   */
+
+  /*
+   * some thoughts: a constraint is a function + lower and upper bounds.  it should support evaluating the constraint, adding it to an optimization problem,
+   * and have support for constraints that require slack variables (adding additional decision variables to the problem).  There
```

Comment:
In the MATLAB implementation of constraint, it does not support adding slack variables through constraints. Instead, the slack variables are added to the optimization problem through `addDecisionVariable` function. I personally prefer to keeping the slack variables in the optimization problems, rather than in the constraint, as one slack variable could be used by several constraints.